src/pki/pkcs.ts

```diff
@@ -0,0 +1,276 @@
+/** based on document Gemeinsame Grundsätze Technik, Anlage 16
+ * (see /docs/documents.md for more info)
+ */
+
+import EncapsulatedContentInfo from "pkijs/src/EncapsulatedContentInfo";
+import SignedData from "pkijs/src/SignedData";
+import SignerInfo from "pkijs/src/SignerInfo";
+import AlgorithmIdentifier from "pkijs/src/AlgorithmIdentifier";
+import IssuerAndSerialNumber from "pkijs/src/IssuerAndSerialNumber";
+import RSASSAPSSParams from "pkijs/src/RSASSAPSSParams";
+import EnvelopedData from "pkijs/src/EnvelopedData";
+import ContentInfo from "pkijs/src/ContentInfo";
+import CertificationRequest from "pkijs/src/CertificationRequest";
+import AttributeTypeAndValue from "pkijs/src/AttributeTypeAndValue";
+import Certificate from "pkijs/src/Certificate";
+import { OctetString, PrintableString, fromBER } from "asn1js";
+import { initCrypto } from "./crypto";
+import { bufferToCertificate, importPKCS8, exportPKCS8, bufferToHex } from "./utils";
+import { File, ResultOrErrors } from "../types";
+import { isIK, isTruncatedIfTooLong, isVarchar, validationByType } from "../validation/utils";
+import { transliterateCertificateName, transliterateRecursively } from "../transcoding";
+
+// - PKCS#7 methods
+
+export const signAndEncryptMessage = async (
+    message: ArrayBuffer,
+    signerCertificate: ArrayBuffer,
+    signerPrivateKey: ArrayBuffer,
+    recipientCertificate: ArrayBuffer
+): Promise<ArrayBuffer> =>
+    await encryptMessage(
+        await signMessage(message, signerCertificate, signerPrivateKey),
+        [recipientCertificate, signerCertificate]
+    );
+
+export const signMessage = async (
+    message: ArrayBuffer,
+    signerCertificate: ArrayBuffer,
+    signerPrivateKey: ArrayBuffer
+): Promise<ArrayBuffer> => {
+    initCrypto();
+
+    const certificate = bufferToCertificate(signerCertificate);
+    const privateKey = await importPKCS8(signerPrivateKey);
+    const signedData = new SignedData({
+        version: 1,
+        encapContentInfo: new EncapsulatedContentInfo({
+            eContentType: "1.2.840.113549.1.7.1", // "data" content type
+            eContent: new OctetString({
+                valueHex: message
+            })
+        }),
+        certificates: [certificate],
+        signerInfos: [
+            new SignerInfo({
+                version: 1,
+                sid: new IssuerAndSerialNumber({
+                    issuer: certificate.issuer,
+                    serialNumber: certificate.serialNumber
+                }),
+            })
+        ]
+    });
+
+    await signedData.sign(privateKey, 0, "SHA-256");
+
+    const cms = new ContentInfo();
+    cms.contentType = "1.2.840.113549.1.7.2"; // signedData
+    cms.content = signedData.toSchema();
+
+    return cms.toSchema().toBER();
```

Comment:
not DER?

Reply:
DER is a subset of BER and it turned out that `toBER` encodes using DER.